modules/social_features/social_embed/src/Controller/EmbedController.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+namespace Drupal\social_embed\Controller;
+
+use Drupal\Component\Uuid\Uuid;
+use Drupal\Core\Ajax\AjaxResponse;
+use Drupal\Core\Ajax\ReplaceCommand;
+use Drupal\Core\Controller\ControllerBase;
+use Drupal\Core\Link;
+use Drupal\social_embed\Service\EmbedHelper;
+use Drupal\url_embed\UrlEmbed;
+use Drupal\user\UserDataInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+/**
+ * Controller for Social Embed endpoint handling.
+ */
+class EmbedController extends ControllerBase {
+
+  /**
+   * The social embed helper services.
+   *
+   * @var \Drupal\social_embed\Service\EmbedHelper
+   */
+  protected EmbedHelper $embedHelper;
+
+  /**
+   * The user data services.
+   *
+   * @var \Drupal\user\UserDataInterface
+   */
+  protected UserDataInterface $userData;
+
+  /**
+   * Url Embed services.
+   *
+   * @var \Drupal\url_embed\UrlEmbed
+   */
+  protected UrlEmbed $urlEmbed;
+
+  /**
+   * The EmbedController constructor.
+   *
+   * @param \Drupal\user\UserDataInterface $user_data
+   *   The user data services.
+   * @param \Drupal\url_embed\UrlEmbed $url_embed
+   *   The url embed services.
+   * @param \Drupal\social_embed\Service\EmbedHelper $embed_helper
+   *   The social embed services.
+   */
+  public function __construct(UserDataInterface $user_data, UrlEmbed $url_embed, EmbedHelper $embed_helper) {
+    $this->userData = $user_data;
+    $this->urlEmbed = $url_embed;
+    $this->embedHelper = $embed_helper;
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @param \Symfony\Component\DependencyInjection\ContainerInterface $container
+   *   The Drupal service container.
+   *
+   * @return static
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('user.data'),
+      $container->get('url_embed'),
+      $container->get('social_embed.helper_service')
+    );
+  }
+
+  /**
+   * Returns the consent of current user.
+   */
+  public function checkConsent(string $provider) {
+    /** @var \Drupal\social_user\Entity\User $account */
+    $account = $this->currentUser();
+    $user_states = $this->userData->get('social_embed', $account->id(), 'consent');
+
+    return $user_states[$provider];
+  }
+
+  /**
+   * Generates embed content of a give URL.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The current request object.
+   *
+   * @return \Drupal\Core\Ajax\AjaxResponse
+   *   The Ajax response.
+   */
+  public function generateEmbed(Request $request) {
+    $url = $request->query->get('url');
+    $uuid = $request->query->get('uuid');
+    if ($url == NULL && Uuid::isValid($uuid) != FALSE) {
+      throw new NotFoundHttpException();
+    }
+    $response = new AjaxResponse();
+    $selector = "#social-embed-iframe-$uuid";
+    if ($info = \Drupal::service('url_embed')->getUrlInfo($url)) {
```

Comment:
doesn't this Throw some exceptions if the provided URL is incorrect? In that case we might want to throw our own?

Reply:
This if else should not be necessary.

If there is nothing to embed so $info isn't set, why do we need the
`      $content = "<div id='social-embed-iframe-$uuid'><p>$link</p></div>";` ? 
We can actually just return the link without our custom markup?